engine/src/main/java/io/camunda/zeebe/engine/processing/deployment/DeploymentCreateProcessor.java

```diff
@@ -131,6 +134,20 @@ public void processRecord(
     }
   }
 
+  @Override
+  public ProcessingError tryHandleError(
+      final TypedRecord<DeploymentRecord> command, final Throwable error) {
+    if (error instanceof IllegalStateException
+        && error.getMessage().contains("this would exceed the maximum batch size")) {
```

Comment:
Just want to mention that there is a new feature right out of the ofen https://github.com/camunda/zeebe/pull/10191 Where you can determine whether the append is successful or not. We can discuss in more detail when I'm back, but in general you could think of whether you would like to handle this either as Remco did via exception or directly on the processing.


Important are these new methods  for you https://github.com/camunda/zeebe/blob/main/engine/src/main/java/io/camunda/zeebe/engine/api/ProcessingResultBuilder.java#L22-L62

Reply:
We should be aware of records that could be too large. I think for now this would only be a risk for deployments and batch activations as you said.
The job batch activate does checks this explicitly here: https://github.com/camunda/zeebe/blob/main/engine/src/main/java/io/camunda/zeebe/engine/processing/job/JobBatchCollector.java#L95.

I think the important thing is that when we write a command rejection, we make sure don't get into an error loop again. This will happens because we receive the "too large" error when processing the rejection as well. Since this is an unexpected error we will handle this and write a rejection because of it, causing the loop here:
https://github.com/camunda/zeebe/blob/main/engine/src/main/java/io/camunda/zeebe/engine/Engine.java#L167-L174

I still have to create a follow-up issue for this, I'll do that later today.

Reply:
Follow-up issue: https://github.com/camunda/zeebe/issues/10199

Reply:
So, if I understand correctly @remcowesterhoud you suggest picking [this](https://github.com/camunda/zeebe/issues/10199) up later and first patch the bug. I think that makes sense.

Personally, I'm not so happy that the DeploymentCreateProcessor is handling an error as expected that it didn't throw. But, that's all good if we make this is a temporary solution and aim to resolve #10199 at a later point.

Did I understand your intention correctly, @remcowesterhoud?

Reply:
@korthout Almost right. The only difference is that I don't see this as a temporary solutions. I intend to keep this code here, next to the changes that will be made in #10199, which I will probably make next week.

I think the misunderstanding here is that the processor will be handling an error that it hasn't thrown. The very first error that triggers the loop is thrown in this processor. The first error happens when the processor tries to append the deployment CREATED event here https://github.com/camunda/zeebe/blob/main/engine/src/main/java/io/camunda/zeebe/engine/processing/deployment/DeploymentCreateProcessor.java#L116.
After this the loop begins and all the consecutive errors are not thrown in this processor, but in the Engine.

In my opinion this is definitely an expected error from a deployment. We know users can upload large processes and we should expect it to happen. Handling it here allows us to give more significant feedback to the user than the generic message of #10199 will do.

Reply:
The problem that I have with this is that the IllegalStateException is thrown by the RecordBatch, which we don't own. The processor now depends on this specific exception being thrown, but it may be changed by the other team. To make sure they don't change it, we need to make sure that this exception is part of the contract when we append records.
 
After discussing F2F, @saig0 @remcowesterhoud and I decided we want to:
- change the exception thrown by RecordBatch into a more meaningful one
- document that throwing this exception is part of the contract with the engine (so DPT doesn't change this); 
- document that this exception can be thrown as part of the contract of the appendX methods of the record writers (for visibility of the exception in the engine processors);
- handle this new exception in the tryHandleError method of the processor (we can later expand this to a separate abstract `handleExceededBatchRecordSize` (or similar) method for further improved visibility in the processors).

All this means that the processor still doesn't throw the exception directly, but it's now part of the contract of appending records. Therefore, the processor is aware that this can happen and it should deal with it accordingly.

Note that we should also update our internal developer handbook on how to write record processors.

Reply:
Please take a look at https://github.com/camunda/zeebe/pull/10191

You can use either the ResultBuilder method which returns an either or which throws an exceptions. As far as I remember  correctly I added it as Javadoc, can't check it right now since I'm underway.

We can also check together tomorrow if you need

Reply:
@Zelldon The problem with using the Either is that we would have to explicitly check the append result for every append we do, since all of the appends could result in this problem. It's easier to have it throw the exception and check it once.

Reply:
Yes I get that just wanted to show the option. Btw I think in some cases it is more likely to fail, multi instance, deployments etc maybe it makes sense to handle here the either just as an idea 🤷‍♂️